lib/IRGen/GenCall.cpp

```diff
@@ -3231,7 +3231,19 @@ class AsyncCallEmission final : public CallEmission {
 
     bool mayReturnErrorDirectly = mayReturnTypedErrorDirectly();
     if (mayReturnErrorDirectly && !nativeSchema.requiresIndirect()) {
-      return emitToUnmappedExplosionWithDirectTypedError(resultType, result,
+      llvm::Value *resultAgg;
+      if (resultTys.size() == 1) {
+        resultAgg = Builder.CreateExtractValue(result, numAsyncContextParams);
+      } else {
+        auto resultTy = llvm::StructType::get(IGM.getLLVMContext(), resultTys);
+        resultAgg = llvm::UndefValue::get(resultTy);
+        for (unsigned i = 0, e = resultTys.size(); i != e; ++i) {
+          llvm::Value *elt =
+              Builder.CreateExtractValue(result, numAsyncContextParams + i);
+          resultAgg = Builder.CreateInsertValue(resultAgg, elt, i);
```

Comment:
It's a little unfortunate to build up an aggregate value only to break it down again, but it works.